### PR TITLE
[Mosaic GPU] Support mixed E4M3/E5M2 FP8 operands in wgmma

### PR DESCRIPTION
**Pull request by mcwitt**
Mixed `e4m3`/`e5m2` matmuls arise in FP8 training; e.g. when gradients are kept in `e5m2` (for dynamic range) while weights/activations are in `e4m3` (for precision), the backward GEMMs pair the two formats (the recipe described in the [FP8-formats paper](https://arxiv.org/abs/2209.05433) and NVIDIA's [Transformer Engine](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/examples/fp8_primer.html)). `wgmma` supports this natively via independent `.atype`/`.btype`, but Pallas/Mosaic previously rejected any operand dtype mismatch.

This PR relaxes the checks that currently require both inputs to have the same element type to allow mixed FP8 types, faithfully reflecting the (FP8) signature of the [wgmma.mma_async](https://docs.nvidia.com/cuda/parallel-thread-execution/#asynchronous-warpgroup-level-matrix-instructions-wgmma-mma) PTX instruction:

<details>

```
FP8 floating point type

wgmma.mma_async.sync.aligned.shape.dtype.atype.btype  d, a-desc, b-desc, scale-d, imm-scale-a, imm-scale-b;

wgmma.mma_async.sync.aligned.shape.dtype.atype.btype  d, a, b-desc, scale-d, imm-scale-a, imm-scale-b;

.shape   = {.m64n8k32, .m64n16k32, .m64n24k32, .m64n32k32,
            .m64n40k32, .m64n48k32, .m64n56k32, .m64n64k32,
            .m64n72k32, .m64n80k32, .m64n88k32, .m64n96k32,
            .m64n104k32, .m64n112k32, .m64n120k32, .m64n128k32,
            .m64n136k32, .m64n144k32, .m64n152k32, .m64n160k32,
            .m64n168k32, .m64n176k32, .m64n184k32, .m64n192k32,
            .m64n200k32, .m64n208k32, .m64n216k32, .m64n224k32,
            .m64n232k32, .m64n240k32, .m64n248k32, .m64n256k32};
.atype  = {.e4m3, .e5m2};
.btype  = {.e4m3, .e5m2};
.dtype  = {.f16, .f32};
```
</details>

### CODE REVIEW

File: jax/_src/pallas/mosaic_gpu/primitives.py [hunk 1]
```diff
@@ -1518,8 +1518,15 @@ def wgmma(acc: gpu_core.WGMMAAbstractAccumulatorRef, a, b) -> None:
         f" rhs={b.shape=}, acc={acc.shape}"
     )
 
-  if a.dtype != b.dtype:
-    raise ValueError(f"Mixed input dtypes for matrix multiplication unsupported: lhs={a.dtype}, rhs={b.dtype}")
+  # A and B must share a dtype, except that the e4m3/e5m2 FP8 pair may be mixed:
+  # `wgmma` takes independent `.atype`/`.btype` operands for FP8.
+  fp8_dtypes = (jnp.float8_e4m3fn, jnp.float8_e5m2)
+  is_mixed_fp8 = a.dtype in fp8_dtypes and b.dtype in fp8_dtypes
```
Comment (apaszke) [inline jax/_src/pallas/mosaic_gpu/primitives.py]: nit: this doesn't check if it's mixed fp8, only if both are fp8. perhaps rename to `both_fp8`?

File: tests/mosaic/gpu_dialect_test.py [hunk 2]
```diff
@@ -651,6 +651,22 @@ def test_wgmma_types_match(self):
     ):
       self.module.operation.verify()
 
+  def test_wgmma_mixed_fp8_operands_are_allowed(self):
```
Comment (apaszke) [inline tests/mosaic/gpu_dialect_test.py:654]: I think this will need a skipTest. The verifier is part of jaxlib and JAX tests must be robust to running with an older jaxlib. Perhaps add a skip unless the jaxlib version (`jtu.jaxlib_version()`) is at least (0, 11, 0)?

File: jax/experimental/mosaic/gpu/wgmma.py [hunk 3]
```diff
@@ -144,14 +148,14 @@ def wgmma_m64(
   if b_k_stride % 16:
     raise ValueError
   # Only 16-bit types support transposes
-  supports_transpose = bytewidth(element_type) == 2
+  supports_transpose = bytewidth(b_element_type) == 2
```
Comment (apaszke) [inline jax/experimental/mosaic/gpu/wgmma.py:152]: nit: could you please assert that a and b element types ahve the same bitwidth?

File: jax/experimental/mosaic/gpu/wgmma.py [hunk 4]
```diff
@@ -353,10 +365,17 @@ def wgmma(
         "WGMMA requires A and B to have the same contraction dimension (K),"
         f" got: {k2} and {k}"
     )
-  if element_type != element_type2:
+  is_mixed_fp8 = (
```
Comment (apaszke) [inline jax/experimental/mosaic/gpu/wgmma.py]: same as before: rename to both_fp8

File: tests/mosaic/gpu_test.py [hunk 5]
```diff
@@ -1234,18 +1290,20 @@ def kernel(ctx, lhs, rhs, out, scratch):
       nvvm.wgmma_wait_group_sync_aligned(0)
       acc.value.store_untiled(out, optimized=False)
 
-    def quantize(x):
+    def quantize(x, exponent_bits, mantissa_bits):
       # Quantize the input to avoid rounding when feeding the WGMMA
       return jax.lax.reduce_precision(x, exponent_bits, mantissa_bits)
```
Comment (apaszke) [inline tests/mosaic/gpu_test.py]: At this point you could simplify this to quantize = jax.lax.reduce_precision

File: tests/pallas/mosaic_gpu_test.py [hunk 6]
```diff
@@ -4541,6 +4541,52 @@ def scope(acc_ref):
         res, a.astype(acc_type) @ b.T.astype(acc_type)
     )
 
+  @parameterized.parameters(
+      (jnp.float8_e4m3fn, jnp.float8_e5m2),
+      (jnp.float8_e5m2, jnp.float8_e4m3fn),
+  )
+  def test_wgmma_mixed_fp8(self, lhs_dtype, rhs_dtype):
+    # PTX `wgmma` takes independent `.atype`/`.btype`, so the e4m3/e5m2 FP8 pair
+    # may be mixed across the two operands.
```
Comment (apaszke) [inline tests/pallas/mosaic_gpu_test.py]: This comment is a bit confusing in this high level test

**Review by apaszke — APPROVED** [review]
Looking great, thank you! But we can't land this without the skip in the test file.

**Comment by apaszke** [comment]
Also: please squash the commits

File: jax/_src/pallas/mosaic_gpu/primitives.py (hunk 1 above)
Comment (mcwitt) [reply jax/_src/pallas/mosaic_gpu/primitives.py]: Good call - fixed in 2a1bfeb6b0a18b981f64babc2a17feb96c29f47e

File: jax/experimental/mosaic/gpu/wgmma.py (hunk 4 above)
Comment (mcwitt) [reply jax/experimental/mosaic/gpu/wgmma.py]: Fixed in 2a1bfeb6b0a18b981f64babc2a17feb96c29f47e

File: tests/mosaic/gpu_dialect_test.py (hunk 2 above)
Comment (mcwitt) [reply tests/mosaic/gpu_dialect_test.py:654]: Added skipTest in 523b6514df3c50f49a4b08e8930ab4395c2a648d

File: jax/experimental/mosaic/gpu/wgmma.py (hunk 3 above)
Comment (mcwitt) [reply jax/experimental/mosaic/gpu/wgmma.py:152]: Added assertion in 15a0aa3b34118131f8bee5c6b18b60b2c994f4af

File: tests/mosaic/gpu_test.py (hunk 5 above)
Comment (mcwitt) [reply tests/mosaic/gpu_test.py]: Removed trivial wrapper in bf13e13

File: tests/pallas/mosaic_gpu_test.py (hunk 6 above)
Comment (mcwitt) [reply tests/pallas/mosaic_gpu_test.py]: Agreed - removed in 08edc18c09d319a8f85e60eace3c4a68f1d100e5

**Review by apaszke — APPROVED** [review]
Thanks!
